cmd/virt-xml-hook/virt_xml_hook.go

```diff
@@ -0,0 +1,139 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/spf13/pflag"
+	"google.golang.org/grpc"
+
+	"kubevirt.io/client-go/log"
+	"kubevirt.io/kubevirt/pkg/hooks"
+	hooksInfo "kubevirt.io/kubevirt/pkg/hooks/info"
+	hooksV1alpha1 "kubevirt.io/kubevirt/pkg/hooks/v1alpha1"
+	hooksV1alpha2 "kubevirt.io/kubevirt/pkg/hooks/v1alpha2"
+)
+
+const (
+	onDefineDomainLoggingMessage = "Hook's OnDefineDomain callback method has been called"
+	usage                        = `updater
+  --version v1alpha1|v1alpha2
+  --args ''`
+)
+
+type infoServer struct {
+	Version string
+}
+
+func (s infoServer) Info(ctx context.Context, params *hooksInfo.InfoParams) (*hooksInfo.InfoResult, error) {
+	log.Log.Info("Hook's Info method has been called")
+
+	return &hooksInfo.InfoResult{
+		Name: "update-xml",
+		Versions: []string{
+			s.Version,
+		},
+		HookPoints: []*hooksInfo.HookPoint{
+			{
+				Name:     hooksInfo.OnDefineDomainHookPointName,
+				Priority: 0,
+			},
+		},
+	}, nil
+}
+
+const virtXML = "virt-xml"
+
+type v1alpha1Server struct {
+	args []string
+}
+
+type v1alpha2Server struct {
+	args []string
+}
+
+func MergeKubeVirtXMLWithProvidedXML(domainXML []byte, args []string) ([]byte, error) {
+	args = append(args, "--edit")
```

Comment:
ok, I'll try it if it works without --edit

Reply:
@andreabolognani  we need either to put `--add-device` or `--edit`, otherwise it fails. Another thing we need to carefully document :/ . I'd assume that a user before trying the sidecar they should manually try with the tool

Reply:
TBH, I wouldn't expect that a user wants to add a device using this sidecar. For this we have hotplugging and you need to have the device in the container. So, it isn't a simple option setting as the case of iotune. I'd prefer to keep it

Reply:
You're right that `--add-device` and `--remove-device` might not be suitable. `--build-xml` clearly doesn't make sense.

But `--edit` on its own means "edit the first instance of the element", and that might not be what the user wants - as evidenced by the fact that you had to use `--edit=all` to your example.

One more thing I just realized. virt-xml can only perform a single change operation at a time, so how would multiple operations look like in the case of hook sidecars? Can you specify the same one with different arguments and expect the XML to go through them in a well-defined order?

If not, then we need to add some looping logic to the hook itself. So maybe the invocation could look like

```yaml
hooks.kubevirt.io/hookSidecars: '[{"args": ["first:disk=iotune.total_bytes_sec=52428800,iotune.total_iops_sec=1000", "all:disk=bus=virtio"], "image": "registry:5000/kubevirt/virt-xml-hook:devel"}]'
```

to mean only the first disk would get iotune parameters, but all of them would have the bus set to "virtio". Those would be translated internally to two separate calls to virt-xml, one with `--edit --disk=iotune.total_bytes_sec=52428800,iotune.total_iops_sec=1000` as arguments and the other one with `--edit=all --disk=bus=virtio`.

Reply:
Yes, I know that it can perform only a single change. This might be enough, again not sure. Today, we have different sidecars for different options, so it will be definitely an improvement. Ok, it is a hack but the `--edit=all` overwrites the first edit, therefore it works. Anyway, after Christmas we can tune this a bit ;)